Add otp email feature for verification

HarzhMehta opened the pull request:
Added OTP-based authentication as an alternative login method. Users can now receive a 6-digit one-time password via email to login instead of using their password.

## Backend Changes
User Model: Added [loginOtp](vscode-file://vscode-app/d:/softwares/VS_Code/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [loginOtpExpires](vscode-file://vscode-app/d:/softwares/VS_Code/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) fields to store OTP data
Auth Routes: Added two new endpoints:
POST /api/auth/request-otp - Generates and emails 6-digit OTP to user
POST /api/auth/verify-otp - Verifies OTP and issues JWT token
Email Service: Integrated Resend mailer for sending OTP emails with HTML template
Environment: Added [RESEND_KEY](vscode-file://vscode-app/d:/softwares/VS_Code/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [RESEND_MAIL](vscode-file://vscode-app/d:/softwares/VS_Code/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) to [.env.example](vscode-file://vscode-app/d:/softwares/VS_Code/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)


## Frontend Changes
Auth Module: Added [requestOtp()](vscode-file://vscode-app/d:/softwares/VS_Code/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [verifyOtp()](vscode-file://vscode-app/d:/softwares/VS_Code/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) functions

## UI Components:
Added OTP input field with 6-digit validation
Added "Request OTP" button
Added toggle link to switch between password and OTP login modes
Popup Logic: Implemented OTP flow handlers with mode switching
Styling: Added secondary button styles for OTP request button

## Security Features
OTP expires after 10 minutes
OTP cleared from database after successful verification (single-use)
OTP fields excluded from default queries
Email validation and user verification before OTP generation

HarzhMehta committed:
add otp email feature for verification

HarzhMehta commented:
I, sorry for delay, can we discuss today - anytime after 3PM / 4PM ?

Harshdev625 commented:
@HarzhMehta Join the Discord: [https://discord.gg/cJvRVDyR](https://discord.gg/cJvRVDyR) and message me whenever you're available. We can discuss there, and then I'll merge your PR.

HarzhMehta commented:
Apologies, we're having trouble coordinating. Since we keep missing each other, could you please review this recording? I've highlighted the files I changed, and the detailed summary is attached. Service used to send emails is "Resend.com"

<img width="785" height="270" alt="image" src="https://github.com/user-attachments/assets/c4041c4e-acb4-4d54-8da5-bc189b9d6526" />


https://github.com/user-attachments/assets/3163e266-6221-414a-b39c-08dd393fbdc2

# Email Verification Flow

## Overview
Email verification with OTP has been implemented. Users must verify their email address during signup before they can login.

## What Changed

### 1. User Model (`models/User.js`)
- Added `isEmailVerified` field (boolean, default: false, indexed)
- Users created via email/password start as unverified
- Google OAuth users are automatically marked as verified

### 2. Signup Flow (`POST /api/auth/signup`)
**Old behavior:** User signs up → immediate JWT token returned → can login
**New behavior:** User signs up → 6-digit OTP sent to email → must verify before login

**Request:**
```json
POST /api/auth/signup
{
  "email": "user@example.com",
  "password": "SecurePass123"
}
```

**Response:**
```json
{
  "message": "Account created successfully. Please check your email for verification code.",
  "email": "user@example.com",
  "requiresVerification": true,
  "expiresIn": 600
}
```

**Email sent:** Subject: "TimeMachine - Verify Your Email" with 6-digit OTP code

### 3. Email Verification (`POST /api/auth/verify-email`) - NEW ENDPOINT
Verifies the OTP and marks the user's email as verified. Returns JWT token on success.

**Request:**
```json
POST /api/auth/verify-email
{
  "email": "user@example.com",
  "otp": "123456"
}
```

**Success Response:**
```json
{
  "message": "Email verified successfully. You can now login.",
  "token": "jwt-token-here",
  "email": "user@example.com",
  "user": {
    "email": "user@example.com",
    "role": "user",
    "timezone": {...},
    "isEmailVerified": true
  }
}
```

**Error Responses:**
- `400` - Email already verified
- `401` - Invalid OTP, expired OTP, or no OTP found
- `500` - Server error

### 4. Resend Verification (`POST /api/auth/resend-verification`) - NEW ENDPOINT
Resends a new OTP if the user didn't receive or lost the first one.

**Request:**
```json
POST /api/auth/resend-verification
{
  "email": "user@example.com"
}
```

**Response:**
```json
{
  "message": "Verification code sent to your email",
  "email": "user@example.com",
  "expiresIn": 600
}
```

### 5. Login Flow (`POST /api/auth/login`)
**New check:** Login now requires email to be verified (except for Google OAuth users)

**Request:**
```json
POST /api/auth/login
{
  "email": "user@example.com",
  "password": "SecurePass123"
}
```

**If email not verified:**
```json
{
  "error": "Email not verified",
  "message": "Please verify your email before logging in. Check your inbox for the verification code.",
  "requiresVerification": true
}
```

**If verified (success):**
```json
{
  "message": "Login successful",
  "token": "jwt-token-here",
  "email": "user@example.com",
  "user": {...}
}
```

### 6. Google OAuth (`POST /api/auth/google`)
Google users are automatically marked as `isEmailVerified: true` since Google has already verified their email.

## Complete User Flow

### New User Registration & Login
```
1. User submits signup form
   ↓
2. POST /api/auth/signup { email, password }
   ↓
3. Account created (isEmailVerified: false)
   ↓
4. 6-digit OTP generated and emailed
   ↓
5. User receives email "TimeMachine - Verify Your Email"
   ↓
6. User enters OTP in verification form
   ↓
7. POST /api/auth/verify-email { email, otp }
   ↓
8. Email marked as verified, JWT token returned
   ↓
9. User can now login with POST /api/auth/login
```

### If User Loses OTP
```
1. User clicks "Resend Code"
   ↓
2. POST /api/auth/resend-verification { email }
   ↓
3. New OTP generated and sent
   ↓
4. User verifies with new OTP
```

## Testing the Flow

### Prerequisites
Ensure `.env` has:
```env
RESEND_KEY=re_YourAPIKeyHere
RESEND_MAIL=onboarding@resend.dev
```

Restart backend after setting these:
```powershell
npm run dev
```

### Test Script (PowerShell)

#### Step 1: Signup (creates unverified user, sends OTP)
```powershell
$signup = Invoke-RestMethod -Method Post -Uri http://localhost:3000/api/auth/signup `
  -ContentType "application/json" `
  -Body '{"email":"test@example.com","password":"TestPass123"}'

$signup
# Should return: requiresVerification: true
```

#### Step 2: Check email for OTP (6-digit code)
Look in the inbox for `test@example.com` for email from `onboarding@resend.dev`

#### Step 3: Verify email with OTP
```powershell
$verify = Invoke-RestMethod -Method Post -Uri http://localhost:3000/api/auth/verify-email `
  -ContentType "application/json" `
  -Body '{"email":"test@example.com","otp":"123456"}'

$verify
# Should return: token and isEmailVerified: true
```

#### Step 4: Try to login (should succeed now)
```powershell
$login = Invoke-RestMethod -Method Post -Uri http://localhost:3000/api/auth/login `
  -ContentType "application/json" `
  -Body '{"email":"test@example.com","password":"TestPass123"}'

$login
# Should return: message: "Login successful", token
```

#### Test 5: Try login BEFORE verification (should fail)
```powershell
# Create another user
Invoke-RestMethod -Method Post -Uri http://localhost:3000/api/auth/signup `
  -ContentType "application/json" `
  -Body '{"email":"unverified@example.com","password":"TestPass123"}'

# Try to login immediately (without verifying)
Invoke-RestMethod -Method Post -Uri http://localhost:3000/api/auth/login `
  -ContentType "application/json" `
  -Body '{"email":"unverified@example.com","password":"TestPass123"}'

# Should return: 403 error "Email not verified"
```

#### Test 6: Resend verification code
```powershell
Invoke-RestMethod -Method Post -Uri http://localhost:3000/api/auth/resend-verification `
  -ContentType "application/json" `
  -Body '{"email":"unverified@example.com"}'

# Should return: "Verification code sent to your email"
```

## API Endpoints Summary

| Endpoint | Method | Auth Required | Purpose |
|----------|--------|---------------|---------|
| `/api/auth/signup` | POST | No | Create account, sends OTP |
| `/api/auth/verify-email` | POST | No | Verify OTP, mark email verified |
| `/api/auth/resend-verification` | POST | No | Resend new OTP |
| `/api/auth/login` | POST | No | Login (requires verified email) |
| `/api/auth/google` | POST | No | Google OAuth (auto-verified) |

## Error Scenarios

### Signup fails to send email
- If `RESEND_KEY` or `RESEND_MAIL` not set: Returns 500 "Failed to send verification email"
- User account is deleted if email send fails
- User must try signup again

### OTP expires
- OTP valid for 10 minutes
- After expiration: Returns 401 "Verification code has expired"
- User must call `/resend-verification` to get new OTP

### Invalid OTP
- Returns 401 "Invalid verification code"
- User can try again with correct OTP
- No limit on retry attempts (but OTP expires after 10 min)

### User tries to verify already-verified email
- Returns 400 "Email is already verified. Please login."

### Login attempt with unverified email
- Returns 403 "Email not verified"
- Response includes `requiresVerification: true` flag
- Frontend should redirect to verification page

## Migration Notes

### Existing Users (Before This Change)
Existing users in the database do NOT have `isEmailVerified` field. MongoDB will treat missing field as `false` by default.

**To allow existing users to login without verification:**
Run this migration script (one-time):

```javascript
// migration-verify-existing-users.js
const mongoose = require('mongoose');
const User = require('./models/User');
require('dotenv').config();

async function migrateExistingUsers() {
  await mongoose.connect(process.env.MONGODB_URI);
  
  const result = await User.updateMany(
    { isEmailVerified: { $exists: false } }, // Users created before this change
    { $set: { isEmailVerified: true } }      // Mark them as verified
  );
  
  console.log(`Updated ${result.modifiedCount} existing users`);
  await mongoose.disconnect();
}

migrateExistingUsers();
```

Run with:
```powershell
node migration-verify-existing-users.js
```

## Security Considerations

1. **OTP Expiry:** 10 minutes (configurable in code)
2. **OTP Format:** 6 digits (100000-999999)
3. **Rate Limiting:** Not implemented (consider adding rate limiting to prevent spam)
4. **OTP Storage:** Stored in plaintext in DB (consider hashing for production)
5. **Email Verification Bypass:** Google OAuth users automatically verified
6. **Account Deletion:** If email send fails during signup, account is deleted (prevents orphaned unverified accounts)

## Frontend Integration Notes

### Signup Flow
1. User fills signup form
2. POST to `/api/auth/signup`
3. On success (status 201):
   - Check response `requiresVerification: true`
   - Redirect to "Verify Email" page
   - Pre-fill email in verification form
4. On error (409): "Email already in use" - show error

### Verification Page
1. Show: "Enter the 6-digit code sent to [email]"
2. Input field for OTP (6 digits)
3. Submit button → POST to `/api/auth/verify-email`
4. "Resend Code" button → POST to `/api/auth/resend-verification`
5. On success: Save JWT token, redirect to dashboard
6. On error: Show error message, allow retry

### Login Flow
1. POST to `/api/auth/login`
2. On success (200): Save token, redirect to dashboard
3. On 403 with `requiresVerification: true`:
   - Redirect to verification page
   - Show message: "Please verify your email first"
   - Provide "Resend Code" option

## Next Steps / Improvements

1. **Add rate limiting** to prevent OTP spam
2. **Hash OTPs** before storing in database
3. **Add email validation** during signup (check MX records)
4. **Implement account lockout** after X failed OTP attempts
5. **Add analytics** to track verification completion rates
6. **Create admin endpoint** to manually verify users if needed
7. **Add "Remember Me"** option for extended token validity
8. **Implement password reset** with similar OTP flow

## Support

If users report not receiving verification emails:
1. Check RESEND_KEY and RESEND_MAIL are set correctly
2. Verify sender domain in Resend dashboard
3. Check Resend dashboard for delivery failures
4. Verify user's email is valid (not typo)
5. Check spam/junk folders
6. Use `/resend-verification` to send new OTP

Harshdev625 commented:
@HarzhMehta You can create an issue for help.  
I think the problem occurs because the login process requires an OTP.  
You’ll need to open an OTP field in the extension and create another route associated with it — that should make it work.
